crates/pessimistic-proof-utils/src/smt.rs

```diff
@@ -0,0 +1,559 @@
+        check_no_duplicates(&kvs);
+        for (key, value) in kvs.iter() {
+            smt.insert(*key, *value).unwrap();
+        }
+        let (key, real_value) = *kvs.choose(&mut rng).unwrap();
+        let proof = smt.get_inclusion_proof(key).unwrap();
+        let fake_value = random();
+        assert_ne!(real_value, fake_value, "Check your rng");
+        assert!(!proof.verify(key, fake_value, smt.root));
+    }
+
+    #[test]
+    fn test_non_inclusion_proof() {
+        let mut rng = thread_rng();
+        let num_keys = rng.gen_range(0..100);
+        let mut smt = Smt::<H, DEPTH>::new();
+        let kvs: Vec<(u32, _)> = (0..num_keys).map(|_| (random(), random())).collect();
+        check_no_duplicates(&kvs);
+        for (key, value) in kvs.iter() {
+            smt.insert(*key, *value).unwrap();
+        }
+        let key: u32 = random();
+        assert!(
+            kvs.iter().position(|(k, _)| k == &key).is_none(),
+            "Check your rng"
+        );
+        let proof = smt.get_non_inclusion_proof(key).unwrap();
+        assert!(proof.verify(key, smt.root, &smt.empty_hash_at_height));
+    }
+
+    #[test]
+    fn test_non_inclusion_proof_failing() {
+        let mut rng = thread_rng();
+        let num_keys = rng.gen_range(1..100);
+        let mut smt = Smt::<H, DEPTH>::new();
+        let kvs: Vec<(u32, _)> = (0..num_keys).map(|_| (random(), random())).collect();
+        check_no_duplicates(&kvs);
+        for (key, value) in kvs.iter() {
+            smt.insert(*key, *value).unwrap();
+        }
+        let (key, _) = *kvs.choose(&mut rng).unwrap();
+        let error = smt.get_non_inclusion_proof(key).unwrap_err();
+        assert_eq!(error, SmtError::KeyPresent);
+    }
+
+    fn test_non_inclusion_proof_and_update(num_keys: usize) {
+        let mut smt = Smt::<H, DEPTH>::new();
+        let kvs: Vec<(u32, _)> = (0..num_keys).map(|_| (random(), random())).collect();
+        check_no_duplicates(&kvs);
+        for (key, value) in kvs.iter() {
+            smt.insert(*key, *value).unwrap();
+        }
+        let key: u32 = random();
+        assert!(
+            kvs.iter().position(|(k, _)| k == &key).is_none(),
+            "Check your rng"
+        );
+        let proof = smt.get_non_inclusion_proof(key).unwrap();
+        assert!(proof.verify(key, smt.root, &smt.empty_hash_at_height));
+        let value = random();
+        let new_root = proof
+            .verify_and_update(key, value, smt.root, &smt.empty_hash_at_height)
+            .unwrap();
+        smt.insert(key, value).unwrap();
+        assert_eq!(smt.root, new_root);
+    }
+
+    #[test]
+    fn test_non_inclusion_proof_and_update_empty() {
+        test_non_inclusion_proof_and_update(0)
+    }
+
+    #[test]
+    fn test_non_inclusion_proof_and_update_nonempty() {
+        let num_keys = thread_rng().gen_range(1..100);
+        test_non_inclusion_proof_and_update(num_keys)
+    }
```

Comment:
With `rstest` you can reduce this to one function: 
```suggestion

    #[rstest]
    #[case(0)]
    #[case(thread_rng().gen_range(1..100))]
    fn test_non_inclusion_proof_and_update(#[case] num_keys: usize) {
        let mut smt = Smt::<H, DEPTH>::new();
        let kvs: Vec<(u32, _)> = (0..num_keys).map(|_| (random(), random())).collect();
        check_no_duplicates(&kvs);
        for (key, value) in kvs.iter() {
            smt.insert(*key, *value).unwrap();
        }
        let key: u32 = random();
        assert!(
            kvs.iter().position(|(k, _)| k == &key).is_none(),
            "Check your rng"
        );
        let proof = smt.get_non_inclusion_proof(key).unwrap();
        assert!(proof.verify(key, smt.root, &smt.empty_hash_at_height));
        let value = random();
        let new_root = proof
            .verify_and_update(key, value, smt.root, &smt.empty_hash_at_height)
            .unwrap();
        smt.insert(key, value).unwrap();
        assert_eq!(smt.root, new_root);
    }
```

Reply:
Nice! @iljakuklic also mentioned this could be cleaned using `proptest/quickcheck`: https://github.com/agglayer/agglayer/pull/163#discussion_r1684061906
I will look into this in a separate PR.